python/lsst/pipe/tasks/setPrimaryFlags.py

```diff
@@ -20,12 +20,153 @@
 # the GNU General Public License along with this program.  If not,
 # see <http://www.lsstcorp.org/LegalNotices/>.
 #
-import numpy
+import numpy as np
 from lsst.pex.config import Config, Field, ListField
 from lsst.pipe.base import Task
 from lsst.geom import Box2D
 
 
+def getPatchInner(sources, patchInfo):
+    """Set a flag for each source if it is in the innerBBox of a patch.
+
+    Parameters
+    ----------
+    sources : `lsst.afw.table.SourceCatalog`
+        A sourceCatalog with pre-calculated centroids.
+    patchInfo : `lsst.skymap.PatchInfo`
+        Information about a `SkyMap` `Patch`.
+
+    Returns
+    --------
+    isPatchInner: boolean array
+        `True` for each source that has a centroid
+        in the inner region of a patch.
+    """
+    # Extract the centroid position for all the sources
+    x = sources["slot_Centroid_x"]
+    y = sources["slot_Centroid_y"]
+    centroidFlag = sources["slot_Centroid_flag"]
+
+    # set inner flags for each source and set primary flags for sources with no children
+    # (or all sources if deblend info not available)
+    innerFloatBBox = Box2D(patchInfo.getInnerBBox())
+    inInner = innerFloatBBox.contains(x, y)
+
+    # When the centroider fails, we can still fall back to the peak, but we don't trust
+    # that quite as much - so we use a slightly smaller box for the patch comparison.
+    shrunkInnerFloatBBox = Box2D(innerFloatBBox)
+    shrunkInnerFloatBBox.grow(-1)
+    inShrunkInner = shrunkInnerFloatBBox.contains(x, y)
+
+    # Flag sources contained in the inner region of a patch
+    isPatchInner = (centroidFlag & inShrunkInner) | (~centroidFlag & inInner)
+    return isPatchInner
+
+
+def getTractInner(sources, tractInfo, skyMap):
+    """Set a flag for each source if it is contained in a given tract.
+
+    Parameters
+    ----------
+    sources : `lsst.afw.table.SourceCatalog`
+        A sourceCatalog with pre-calculated centroids.
+        If `key` is not `None` then the results are written into
+        `sources[key]`.
+    tractInfo : `lsst.skymap.TractInfo`
+        Tract object
+    skyMap : `lsst.skymap.BaseSkyMap`
+        Sky tessellation object
+
+    Returns
+    -------
+    isTractInner: `bool` array
+        True for each source that has a centroid in the same
+        tract as `tractInfo`.
+    """
+    tractId = tractInfo.getId()
+    isTractInner = np.array([skyMap.findTract(s.getCoord()).getId() == tractId for s in sources])
+    return isTractInner
+
+
+def getPseudoSources(sources, pseudoFilterList, schema, log):
+    """Get a flag that marks pseudo sources.
+
+    Some categories of sources, for example sky objects,
+    are not really detected sources and should not be considered primary
+    sources.
```

Comment:
I'm having trouble trying to find a pattern on your choices for line-breaks.  You don't seem to ever really use the full 79 chars, and sometimes truncate differently in the same set of lines (as above!).

Reply:
The problem with having a line limit for docstrings and code comments is that if you write all the way to the end of the line and want to edit it later, you have to reformat the entire paragraph after the edit, even if you're a single character or two over the very small line limit. So I always give myself a large buffer and try to break at logical points where there might be an edit later, for example a comma or period. But at the end of the paragraph, if there are only a few words on the next line, I don't really care as much about leaving space so I tend to go write all the way to the end.

It's why I'm a huge supporter of a suggested line length for docstrings and comments but no strict enforcement, which keeps the diffs down and makes it easier (and thus more likely) that I'll keep docs up to date.